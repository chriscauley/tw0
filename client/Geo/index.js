const { range } = require('lodash')
const alphanum = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'.split(
  '',
)
const numalpha = '0123456789abcdefghijklmnopqrstuvqxwyz'.split('')

const mod = (n, d) => ((n % d) + d) % d

const Look = (geo) => {
  const make = (shape, dist) => {
    if (dist > 0 && !look[shape][dist - 1]) {
      make(shape, dist - 1)
    }
    if (shape.startsWith('__')) {
      // dunder means "outer shell of look"
      look[shape][dist] = shapes[shape.slice(2)](dist)
    } else {
      if (!look['__' + shape][dist]) {
        // need outer shells to make filled looks
        make('__' + shape, dist)
      }
      look[shape][dist] = []
      range(dist + 1).forEach((_dist) => {
        look[shape][dist] = look[shape][dist].concat(look['__' + shape][_dist])
      })
    }
  }

  const shapes = {
    f: (dist) => {
    },
    box: (dist) => {
      const out = []
      geo.dindexes.forEach((dindex, i) => {
        const o_dindex = geo.dindexes[(i + 1) % geo.dindexes.length] // orthogonal direction
        range(dist * 2 + 1).forEach((o_dist) => {
          out.push(dist * dindex + o_dist * o_dindex)
        })
        out.pop() // last corner is repeated
      })
      return out
    },
    // "up right down left"
    urdl: (dist) => geo.dindexes.map((di) => di * dist),
    circle: (dist) => {
      const out = []
      geo.dindexes.forEach((dindex, i) => {
        out.push(dindex * dist)
        const o_dindex = geo.dindexes[(i + 1) % geo.dindexes.length] // orthogonal direction
        range(1, dist).forEach((o_dist) =>
          out.push((dist - o_dist) * dindex + o_dist * o_dindex),
        )
      })
      return out
    },
  }

  const look = (shape, index, dist) => {
    if (!look[shape][dist]) {
      make(shape, dist)
    }
    return look[shape][dist].map((dindex) => index + dindex)
  }

  Object.keys(shapes).forEach((shape) => {
    look[shape] = [[]] // all geometries only see nothing at range 0
    look['__' + shape] = [[]]
    make(shape, 1)
  })

  look.inBounds = (shape, index, dist) => {
    // like look, but won't cross out of bounds
    const x0 = geo.i2xy(index)[0]
    const xys = look(shape, index, dist)
      .map((i) => geo.i2xy(i))
      .filter(
        (xy) =>
          Math.abs(xy[0] - x0) <= dist &&
          xy[1] < geo.H &&
          xy[1] >= geo.y0 &&
          xy[1] < geo.y0 + geo.H,
      )
    return xys.map((xy) => geo.xy2i(xy))
  }

  return look
}

const Geo = (x0, x_max, y0, y_max) => {
  if (y0 === undefined) {
    y_max = x_max - 1
    x_max = x0 - 1
    x0 = 0
    y0 = 0
  }

  const W = Math.abs(x_max - x0) + 1
  const H = Math.abs(y_max - y0) + 1

  const geo = {
    x0,
    y0,
    W,
    H,
    xys: [],
    indexes: [],
    AREA: W * H,
    dindexes: [-W, 1, W, -1],
    i2xy: (i) => [mod(i, W), Math.floor(i / W)],
    xy2i: (xy) => xy[0] + xy[1] * W,
    print(board, {from_xy=[x0,y0], to_xy=[x_max, y_max], delimiter='',empty=' ', extras}={}) {
      const xs = range(from_xy[0], to_xy[0]+1)
      const ys = range(from_xy[1], to_xy[1]+1)
      const lines = ys.map((y) => (
        xs.map((x) => board[this.xy2i([x, y])])
          .map(s => s=== undefined ? empty : s)
          .join(delimiter)
      ))
      if (extras) {
        extras.forEach((e, i) => lines[i].push('  ' + e))
      }
      return lines.join('\n')
    },
    log: (board, options) => console.log(geo.print(board, options)),
    makeBoard(xys, values = []) {
      const board = {}
      xys.forEach((xy, i) => {
        board[this.xy2i(xy)] = values[i] || i
      })
      return board
    },
    inBounds(xy) {
      return xy[0] >= x0 && xy[0] < x0 + W && xy[1] >= y0 && xy[1] < y0 + H
    },
    eachXY(f) {
      console.warn('eachXY is depracated, use geo.xys.forEach instead')
      range(x0, x_max+1).forEach(
        x => range(y0, y_max+1).forEach(
          y => f([x,y])
        )
      )
    },
    slice(board, xy, W, H, _default) {
      const out = []
      const ys = range(xy[1], xy[1] + H)
      range(xy[0], xy[0]+W).forEach(
        x => ys.forEach(
          y => {
            const v = board[geo.xy2i([x,y])]
            out.push(v === undefined ? _default : v)
          }
        )
      )
      return out
    },
    getDindexes: (dindex) => geo.rot_dindexes[dindex],
    rot_dindexes: {},
  }

  geo.CENTER = geo.xy2i([Math.floor((geo.x0+geo.W)/2), Math.floor((geo.y0+geo.H)/2)])

  const _double_dindexes = geo.dindexes.concat(geo.dindexes)
  geo.dindexes.forEach((dindex, i) => {
    geo.rot_dindexes[dindex] = _double_dindexes.slice(i, i+geo.dindexes.length)
  })

  range(x0, x_max+1).forEach(
    x => range(y0, y_max+1).forEach(
      y => {
        const xy = [x,y]
        geo.xys.push()
        geo.indexes.push(geo.xy2i(xy))
      }
    )
  )

  geo.look = Look(geo)
  return geo
}
const assert = (bool, exception) => {
  if (!bool) {
    throw exception
  }
}

const log = (...args) => {
  if (log.isOn()) {
    console.log(...args)
  }
}

log.isOn = () => process.argv.includes('-v')

const answer = (text, value, expected) => {
  if (expected !== undefined) {
    assert(value === expected, `Bad answer for ${text}, ${value} !== ${expected}`)
  }
  console.log(text, value)
}

// eslint-disable-next-line
export default Geo
//   numalpha,
//   alphanum,
//   alphabet: alphanum.slice(0,26),
//   Geo,
//   assert,
//   log,
//   answer,
//   mod,
// }
