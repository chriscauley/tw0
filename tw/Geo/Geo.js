import { range } from 'lodash'

import { mod } from '../utils'
import Look from './Look'

const geo_cache = {}

export default (x0, x_max, y0, y_max) => {
  const key = `${x0},${x_max},${y0},${y_max}`
  if (geo_cache[key]) {
    return geo_cache[key]
  }
  if (x_max === undefined) {
    x_max = x0
  }
  if (y0 === undefined) {
    y_max = x_max - 1
    x_max = x0 - 1
    x0 = 0
    y0 = 0
  }

  const W = Math.abs(x_max - x0) + 1
  const H = Math.abs(y_max - y0) + 1

  const geo = (geo_cache[key] = {
    x0,
    y0,
    W,
    H,
    xys: [],
    indexes: [],
    AREA: W * H,
    _dindex_names: ['u', 'l', 'r', 'd'],
    _dindex_chars: ['^', '<', '>', 'v'],
    _dindex2char: {},
    _name2dindex: {},
    _dindex2name: {},
    dindexes: [-W, -1, 1, W], // u, l, r, d
    rot_dindexes: {
      [-W]: [-W, -1, 1, W], // u, l, r, d
      [W]: [W, 1, -1, -W], // d, l, r, u
      [1]: [1, -W, W, -1], // r, u, d, l
      [-1]: [-1, W, -W, 1], // l, d, u, r
    },
    index2xy: (i) => [mod(i, W), Math.floor(i / W)],
    xy2index: (xy) => mod(xy[0] + xy[1] * W, geo.AREA),
    print(
      board,
      {
        from_xy = [x0, y0],
        to_xy = [x_max, y_max],
        delimiter = '',
        empty = ' ',
        extras,
        title,
      } = {},
    ) {
      const xs = range(from_xy[0], to_xy[0] + 1)
      const ys = range(from_xy[1], to_xy[1] + 1)
      const lines = ys.map((y) =>
        xs
          .map((x) => board[this.xy2index([x, y])])
          .map((s) => (s === undefined ? empty : s))
          .join(delimiter),
      )
      if (extras) {
        extras.forEach((e, i) => (lines[i] += '\t' + e))
      }
      title && lines.unshift(title)
      return lines.join('\n')
    },
    inBounds(xy) {
      return xy[0] >= x0 && xy[0] < x0 + W && xy[1] >= y0 && xy[1] < y0 + H
    },
    slice(xy, W, H) {
      const out = []
      const ys = range(xy[1], xy[1] + H)
      const xs = range(xy[0], xy[0] + W)
      ys.forEach((y) => xs.forEach((x) => out.push(geo.xy2index([x, y]))))
      return out
    },
    floorDindex(dindex) {
      // if dindex is a multiple of W it is in thy y direction
      // otherwise it is in the x direction
      return (dindex % W === 0 ? W : 1) * Math.sign(dindex)
    },
  })

  geo.dindexes.forEach((dindex, i) => {
    const name = geo._dindex_names[i]
    geo._dindex2name[dindex] = name
    geo._name2dindex[name] = dindex
    geo._dindex2char[dindex] = geo._dindex_chars[i]
  })

  geo.CENTER = geo.xy2index([Math.floor((geo.x0 + geo.W) / 2), Math.floor((geo.y0 + geo.H) / 2)])

  range(y0, y_max + 1).forEach((y) =>
    range(x0, x_max + 1).forEach((x) => {
      const xy = [x, y]
      geo.xys.push(xy)
      geo.indexes.push(geo.xy2index(xy))
    }),
  )

  geo.look = Look(geo)
  return geo
}
