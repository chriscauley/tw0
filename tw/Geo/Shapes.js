import { range } from 'lodash'

const Shapes = (geo) => ({
  circle: (dist, dindex) => {
    const out = []
    const o_dindex = geo.rot_dindexes[dindex][1]
    out.push(dindex * dist)
    range(1, dist + 1).forEach((i) => {
      out.push(dindex * (dist - i) - o_dindex * i)
      out.push(dindex * (dist - i) + o_dindex * i)
    })
    range(1, dist).forEach((i) => {
      out.push(-dindex * (dist - i) - o_dindex * i)
      out.push(-dindex * (dist - i) + o_dindex * i)
    })
    out.push(dindex * -dist)
    return out
  },
  box: (dist, dindex) => {
    const out = []
    const o_dindex = geo.rot_dindexes[dindex][1]

    // top row
    range(-dist, dist + 1).forEach((i) => out.push(dist * dindex - i * o_dindex))

    // left and right sides in middle
    range(1, 2 * dist).forEach((i) => {
      out.push((dist - i) * dindex + o_dindex * dist)
      out.push((dist - i) * dindex - o_dindex * dist)
    })

    // bottom row
    range(-dist, dist + 1).forEach((i) => out.push(-dist * dindex - i * o_dindex))
    return out
  },
  three: (dist, dindex) => {
    const target = dist * dindex
    const o_dindex = geo.rot_dindexes[dindex][1]
    return [target, target + o_dindex, target - o_dindex]
  },
  cone: (dist, dindex) => {
    const target = dist * dindex
    const out = [target]
    const o_dindex = geo.rot_dindexes[dindex][1]
    range(1, dist).forEach((o_dist) => {
      out.push(target + o_dist * o_dindex)
      out.push(target - o_dist * o_dindex)
    })
    return out
  },
})

Shapes.list = []

Object.keys(Shapes()).forEach((shape) => {
  Shapes.list.push('__' + shape)
  Shapes.list.push(shape)
})

export default Shapes
