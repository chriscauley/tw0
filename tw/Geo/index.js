const { range } = require('lodash')
const alpha = 'abcdefghijklmnopqrstuvwxyz'
const ALPHA = alpha.toUpperCase()
const numbers = '0123456789'
export const alphabet = alpha.split('')
export const alphanum = (alpha+ALPHA+numbers).split('')
export const numalpha = (numbers+alpha+ALPHA).split('')

const mod = (n, d) => ((n % d) + d) % d
export const SHAPES = []

const Look = (geo) => {
  const make = (shape, dist) => {
    const R = 1
    if (look[shape][R][dist]) {
      return
    }
    if (dist > 0 && !look[shape][R][dist - 1]) {
      make(shape, dist - 1)
    }
    if (shape.startsWith('__')) {
      // dunder means "outer shell of look"
      geo.dindexes.forEach(dindex => {
        look[shape][dindex][dist] = shapes[shape.slice(2)](dist, dindex)
      })
    } else {
      if (!look['__' + shape][R][dist]) {
        // need outer shells to make filled looks
        make('__' + shape, dist)
      }
      geo.dindexes.forEach(dindex => {
        look[shape][dindex][dist] = []
        range(dist + 1).forEach((_dist) => {
          look[shape][dindex][dist] = look[shape][dindex][dist].concat(look['__' + shape][dindex][_dist])
        })
      })
    }
  }

  const shapes = {
    box: (dist, dindex) => {
      const out = []
      const o_dindex = geo.rot_dindexes[dindex][1]

      // top row
      range(-dist, dist+1).forEach(i => out.push(dist * dindex - i * o_dindex))

      // left and right sides in middle
      range(1, 2*dist).forEach(i => {
        out.push((dist-i) * dindex + o_dindex * dist)
        out.push((dist-i) * dindex - o_dindex * dist)
      })

      // bottom row
      range(-dist, dist+1).forEach(i => out.push(-dist * dindex - i * o_dindex))
      return out
    },
    circle: (dist, dindex) => {
      const out = []
      const o_dindex = geo.rot_dindexes[dindex][1]
      out.push(dindex * dist)
      range(1, dist+1).forEach(i => {
        out.push(dindex*(dist-i) - o_dindex * i)
        out.push(dindex*(dist-i) + o_dindex * i)
      })
      range(1, dist).forEach(i => {
        out.push(-dindex*(dist-i) - o_dindex * i)
        out.push(-dindex*(dist-i) + o_dindex * i)
      })
      out.push(dindex * -dist)
      return out
    },
  }

  const look = (shape, index, dist, dindex) => {
    if (!look[shape][dindex]) {
      throw Error(`Invalid dindex: ${dindex}`)
    }
    if (!look[shape][dindex][dist]) {
      make(shape, dist)
    }
    return look[shape][dindex][dist].map((dindex) => index + dindex)
  }

  const registerShape = (name) => {
    SHAPES.push(name)
    look[name] = {}
    geo.dindexes.forEach(dindex => {
      look[name][dindex] = [[]] // all geometries only see nothing at range 0
    })
  }

  Object.keys(shapes).forEach((shape) => {
    registerShape(shape)
    registerShape('__' + shape)
    make(shape, 1)
  })

  look.inBounds = (shape, index, dist) => {
    // like look, but won't cross out of bounds
    const x0 = geo.index2xy(index)[0]
    const xys = look(shape, index, dist)
      .map((i) => geo.index2xy(i))
      .filter(
        (xy) =>
          Math.abs(xy[0] - x0) <= dist &&
          xy[1] < geo.H &&
          xy[1] >= geo.y0 &&
          xy[1] < geo.y0 + geo.H,
      )
    return xys.map((xy) => geo.xy2index(xy))
  }

  return look
}

const Geo = (x0, x_max, y0, y_max) => {
  if (x_max === undefined) {
    x_max = x0
  }
  if (y0 === undefined) {
    y_max = x_max - 1
    x_max = x0 - 1
    x0 = 0
    y0 = 0
  }

  const W = Math.abs(x_max - x0) + 1
  const H = Math.abs(y_max - y0) + 1

  const geo = {
    x0,
    y0,
    W,
    H,
    xys: [],
    indexes: [],
    AREA: W * H,
    _dindex_names: ['u', 'l', 'r', 'd',],
    _dindex_chars: ['^', '<', '>', 'v',],
    _dindex2char: {},
    _name2dindex: {},
    _dindex2name: {},
    dindexes: [-W, -1, 1, W ], // u, l, r, d
    rot_dindexes: {
      [-W]: [-W, -1, 1, W ], // u, l, r, d
      [W]: [W, 1, -1, -W ], // d, l, r, u
      [1]: [1, -W, W, -1], // r, u, d, l
      [-1]: [-1, W, -W, 1], // l, d, u, r
    },
    index2xy: (i) => [mod(i, W), Math.floor(i / W)],
    xy2index: (xy) => mod(xy[0] + xy[1] * W, geo.AREA),
    print(board, {from_xy=[x0,y0], to_xy=[x_max, y_max], delimiter='',empty=' ', extras, title}={}) {
      const xs = range(from_xy[0], to_xy[0]+1)
      const ys = range(from_xy[1], to_xy[1]+1)
      const lines = ys.map((y) => (
        xs.map((x) => board[this.xy2index([x, y])])
          .map(s => s=== undefined ? empty : s)
          .join(delimiter)
      ))
      if (extras) {
        extras.forEach((e, i) => lines[i] += '\t' + e)
      }
      title && lines.unshift(title)
      return lines.join('\n')
    },
    log: (board, options) => console.log(geo.print(board, options)),
    inBounds(xy) {
      return xy[0] >= x0 && xy[0] < x0 + W && xy[1] >= y0 && xy[1] < y0 + H
    },
    slice(xy, W, H) {
      const out = []
      const ys = range(xy[1], xy[1] + H)
      const xs = range(xy[0], xy[0]+W)
      ys.forEach(
        y => xs.forEach(
          x=> out.push(geo.xy2index([x,y]))
        )
      )
      return out
    },
  }

  geo.dindexes.forEach((dindex, i) => {
    const name = geo._dindex_names[i]
    geo._dindex2name[dindex] = name
    geo._name2dindex[name] = dindex
    geo._dindex2char[dindex] = geo._dindex_chars[i]
  })

  geo.CENTER = geo.xy2index([Math.floor((geo.x0+geo.W)/2), Math.floor((geo.y0+geo.H)/2)])

  range(y0, y_max+1).forEach(
    y => range(x0, x_max+1).forEach(
      x => {
        const xy = [x,y]
        geo.xys.push()
        geo.indexes.push(geo.xy2index(xy))
      }
    )
  )

  geo.look = Look(geo)
  return geo
}
const assert = (bool, exception) => {
  if (!bool) {
    throw exception
  }
}

const log = (...args) => {
  if (log.isOn()) {
    console.log(...args)
  }
}

log.isOn = () => process.argv.includes('-v')

const answer = (text, value, expected) => {
  if (expected !== undefined) {
    assert(value === expected, `Bad answer for ${text}, ${value} !== ${expected}`)
  }
  console.log(text, value)
}

// eslint-disable-next-line
export default Geo
//   numalpha,
//   alphanum,
//   alphabet: alphanum.slice(0,26),
//   Geo,
//   assert,
//   log,
//   answer,
//   mod,
// }
