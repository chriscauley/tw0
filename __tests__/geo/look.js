import Geo from '../../client/Geo'

const snapIndexes = (geo, indexes) => {
  let error
  try {
    expect(indexes.join(',')).toMatchSnapshot()
  } catch (e) {
    error = e
  }
  if (process.argv.includes('-V')) {
    const board = {}
    board[geo.CENTER] = 'X'
    indexes.forEach((index, i) => board[index] = i)
    geo.log(board, {delimiter:'.'})
  }
  if (error) { throw error }
}

test('`geo.look("circle", index, dist, dindex)` changes direction with dindex', () => {
  const geo = Geo(5,5)
  const board = {}
  const index = geo.CENTER
  geo.dindexes.forEach(dindex=> {
    const indexes = geo.look('circle', index, 1, dindex)
    snapIndexes(geo, indexes)
  })
})