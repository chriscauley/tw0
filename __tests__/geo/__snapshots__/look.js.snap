// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`geo.look("circle", index, dist, dindex)\` changes direction with dindex 1`] = `"7,13,17,11"`;

exports[`\`geo.look("circle", index, dist, dindex)\` changes direction with dindex 2`] = `"7,13,17,11"`;

exports[`\`geo.look("circle", index, dist, dindex)\` changes direction with dindex 3`] = `"7,13,17,11"`;

exports[`\`geo.look("circle", index, dist, dindex)\` changes direction with dindex 4`] = `"7,13,17,11"`;
